File: src/main/java/jvsk/java/Maybe.java
package jvsk.java;

public sealed interface Maybe<T> {

    record Just<T>(T value) implements Maybe<T> {}

    final class None<T> implements Maybe<T> {

        @Override
        public String toString() {
            return "None";
        }
    }

    static <T> Just<? extends T> just(final T value) {
        return new Just<>(value);
    }

    static <T> None<? extends T> none() {
        return new None<>();
    }
}

File: src/main/kotlin/jvsk/kotlin/Maybe.kt
package jvsk.kotlin

sealed interface Maybe<out T> {

    data class Just<T>(val value: T) : Maybe<T>

    data object None : Maybe<Nothing>
}

fun CharSequence.findChar(c: Char): Maybe<Int> {
    for (i in indices) {
        if (get(i) == c) {
            return Maybe.Just(i)
        }
    }
    return Maybe.None
}

fun firstSegment(string: CharSequence): CharSequence =
    when (val index = string.findChar(':')) {
        is Maybe.Just -> string.subSequence(0, index.value)
        Maybe.None -> string
    }
